### ISSUE TITLE
rm doesn't really traverse directories like GNU rm at all, will fail on deeply-nested paths

### ISSUE THREAD

**Issue by untitaker** [uutils/coreutils#2949]
GNU `rm` avoids constructing path strings of unbounded length to allow recursively deleting directories whose absolute paths exceed the max path length a Linux syscall can take. For more information please refer to this blogpost I wrote a while ago: https://unterwaditzer.net/2021/linux-paths.html

The `rm` in this repo appears to construct `Path` objects and therefore almost certainly will fail to remove the nested directories constructed in that article.

IIRC BSD `rm` works similarly to GNU `rm` here, and I would not be surprised if POSIX mandates a solution that works on really deeply nested filepaths.

There's probably more utilities in this repo which use `Path` objects, which immediately makes them fail in those edgecase situations where the equivalent GNU version doesn't.

**Comment by tertsdiepraam** [comment]
Interesting! So, if I understand correctly GNU does a sort of depth-first-traversal of the directories, removing each one it steps out of? We're currently traversing the directories with `WalkDir`. They have an [open issue](https://github.com/BurntSushi/walkdir/issues/23) for this too, with some interesting discussion.

**Comment by untitaker** [comment]
that's exactly the issue. i attempted to implement directory traversal with openat here: https://github.com/untitaker/fdwalk

it will depend on usecase whether that's feasible. for ripgrep i believe it doesn't make any sense, as it needs to print filepaths, keep track of globs, ignorefiles, etc. for something "simple" and "essential" like coreutils i think it's necessary. it's also not too hard to do it as long as you don't use verbose option.

**Comment by tavianator** [comment]
> I would not be surprised if POSIX mandates a solution that works on really deeply nested filepaths.

Indeed, https://pubs.opengroup.org/onlinepubs/9699919799/utilities/rm.html says:

> The *rm* utility shall be able to descend to arbitrary depths in a file hierarchy, and shall not fail due to path length limitations (unless an operand specified by the user exceeds system limitations).

**Comment by untitaker** [comment]
this is still an issue.

**Comment by valpackett** [comment]
Please consider [cap-std](https://github.com/bytecodealliance/cap-std) for working with nested paths.

**Comment by untitaker** [comment]
this issue might actually be fixed by https://github.com/uutils/coreutils/pull/8517

> Please consider [cap-std](https://github.com/bytecodealliance/cap-std?rgh-link-date=2025-09-17T18%3A16%3A28.000Z) for working with nested paths.

if you want to recursively delete a directory, you will eventually want to run `openat(X, "../")` to exit the recursion, and then delete that directory. from what i understand cap-std is designed to prevent this. keeping the parent file descriptor open will cause another pathological case where you run out of file descriptors in deeply-nested directories

**Comment by the8472** [comment]
> openat(X, "../")

Note that you have to be extremely careful around this, otherwise you expose yourself confused deputy attacks.
See the discussion in https://github.com/rust-lang/rust/pull/93160#discussion_r789702060
